### DISCUSSION TITLE
allow plugin to do url auth with body and headers

### DISCUSSION THREAD

**Discussion by yosiasz** [grafana/grafana-infinity-datasource#323] [Ideas]
Greetings,

working with centreon rest api. The feature I would like to request is add ability to use an authentication url in the list of authentication choices

![image](https://user-images.githubusercontent.com/900951/172890828-a51f5665-224b-4caf-8866-ab232e710d6a.png)

https://docs.centreon.com/docs/api/rest-api-v1/#authentication

Unless there already is a way of doing that.

Amazing plugin!

Thanks

**Comment by yesoreyeram** [comment]
Hi. I am not sure if there is any standard around this. Bit tricky to support this. There are few outstanding questions..

* What value the authenticate request returns? How to parse the response? What to extract and where to pass? ( like headers, bearer token etc )..
* How frequently the token timeouts before retry.
* How to differentiate the token expiry and invalid token? ( for retry purposes )

Happy to implement a generic authentication mechanism if we found clarity around this.  Doesn't make any sense designing this feature just to target centreon.

**Reply by yosiasz** [reply]
> Hi. I am not sure if there is any standard around this. Bit tricky to support this. There are few outstanding questions..
> 
> * What value the authenticate request returns? How to parse the response? What to extract and where to pass? ( like headers, bearer token etc )..

Response is
{
    "authToken": "NTc1MDU3MGE3M2JiODIuMjA4OTA2OTc="
}
> * How frequently the token timeouts before retry.

Checking with centreon

> * How to differentiate the token expiry and invalid token? ( for retry purposes )
> 
Good question, will ask centreon
> Happy to implement a generic authentication mechanism if we found clarity around this. Doesn't make any sense designing this feature just to target centreon.


I agree! Maybe they need to get something more robust

Thanks

**Reply by yosiasz** [reply]
for now I just wrote my own node api to do the work for me, plug it into a grafana variable with inifinity plugin and it works

```
const form = new FormData();
    form.append("username", 'jazz'); //fake :)
    form.append('password', 'circleof5th'); //fake :)
    try {
      const url = 'http://your.centreon.ip.or.FQDN/centreon/api/index.php?action=authenticate';
      const response = await fetch(url, {method: 'POST', body: form});     
      const data = await response.json() 
      return res.json(data);
    } catch (e) {
      throw e.response ? e.response.body.message : e;
    }
```

**Reply by yosiasz** [reply]
They say about an hour ttl. They dont provide a timestamp value unfortunately

**Reply by rootolo** [reply]
Hi,
could you describe with more details how to configure infinity plugin to use it with Centreon ?

Thanks a million
   Davide


> for now I just wrote my own node api to do the work for me, plug it into a grafana variable with inifinity plugin and it works
> 
> ```
> const form = new FormData();
>     form.append("username", 'jazz'); //fake :)
>     form.append('password', 'circleof5th'); //fake :)
>     try {
>       const url = 'http://your.centreon.ip.or.FQDN/centreon/api/index.php?action=authenticate';
>       const response = await fetch(url, {method: 'POST', body: form});     
>       const data = await response.json() 
>       return res.json(data);
>     } catch (e) {
>       throw e.response ? e.response.body.message : e;
>     }
> ```
